Libplanet/Blockchain/BlockChain.cs

```diff
@@ -917,13 +917,22 @@ public long GetNextTxNonce(Address address)
                 try
                 {
                     SetStates(block, actionEvaluations, false);
+                    block = new Block<T>(block, trieStateStore.GetRootHash(block.Hash));
+
+                    // it's needed because `block.Hash` was updated with the state root hash.
+                    // FIXME: we need a method for calculating the state root hash without
+                    // `.SetStates()`.
+                    SetStates(block, actionEvaluations, false);
```

Comment:
It seems some duplicated action is conducted in here. Can't we reuse some evaluations from first `SetStates()`?

Reply:
`.SetState()` doesn't return any value. it only stores to `.StateStore`. 

We need the state root hash to create a new block. and, at this point, the only way to do this is to get the stored state root hash via `.SetState()`.

Reply:
I mean
```
            IImmutableSet<Address> stateUpdatedAddresses = actionEvaluations
                .SelectMany(a => a.OutputStates.StateUpdatedAddresses)
                .ToImmutableHashSet();
            IImmutableSet<(Address, Currency)> updatedFungibleAssets = actionEvaluations
                .SelectMany(a => a.OutputStates.UpdatedFungibleAssets
                    .SelectMany(kv => kv.Value.Select(c => (kv.Key, c))))
                .ToImmutableHashSet();
```
reuse these values which are inside of `SetStates()`.

Reply:
Ah, I see.

Reply:
@limebell On second thought, if we adjust codes as above `FIXME:`, we can omit to repeat `SetStates()` calls and it would be more natural to me. so I'll waive [it](https://github.com/planetarium/libplanet/pull/1116#discussion_r535806980) at this time.